contracts/Marketplace.sol

```diff
@@ -149,28 +149,47 @@ contract Marketplace is SlotReservations, Proofs, StateRetrieval, Endian {
     slot.slotIndex = slotIndex;
     RequestContext storage context = _requestContexts[requestId];
 
-    require(slotState(slotId) == SlotState.Free, "Slot is not free");
+    require(
+      slotState(slotId) == SlotState.Free ||
+        slotState(slotId) == SlotState.Repair,
+      "Slot is not free"
+    );
 
     _startRequiringProofs(slotId, request.ask.proofProbability);
     submitProof(slotId, proof);
 
     slot.host = msg.sender;
-    slot.state = SlotState.Filled;
     slot.filledAt = block.timestamp;
 
     context.slotsFilled += 1;
     context.fundsToReturnToClient -= _slotPayout(requestId, slot.filledAt);
 
     // Collect collateral
-    uint256 collateralAmount = request.ask.collateral;
+    uint256 collateralAmount;
+    if (slotState(slotId) == SlotState.Repair) {
+      // Host is repairing a slot and is entitled for repair reward, so he gets "discounted collateral"
+      // in this way he gets "physically" the reward at the end of the request when the full amount of collateral
+      // is returned to him.
+      collateralAmount =
+        request.ask.collateral -
+        ((request.ask.collateral * _config.collateral.repairRewardPercentage) /
+          100);
+    } else {
+      collateralAmount = request.ask.collateral;
+    }
     _transferFrom(msg.sender, collateralAmount);
     _marketplaceTotals.received += collateralAmount;
-    slot.currentCollateral = collateralAmount;
+    slot.currentCollateral = request.ask.collateral; // Even if he has collateral discounted, he is operating with full collateral
 
     _addToMySlots(slot.host, slotId);
 
+    slot.state = SlotState.Filled;
     emit SlotFilled(requestId, slotIndex);
-    if (context.slotsFilled == request.ask.slots) {
+
+    if (
+      context.slotsFilled == request.ask.slots &&
+      context.state != RequestState.Started // In case of repair, we already have the request started, so we prevent from "restarting it"
+    ) {
```

Comment:
What is this event being used for in the case of repair? 

Emitting this event in the case of repair will cause Codex to re-add all the slots in the request. Currently, Codex already handles the `SlotFreed` event, re-adding that slot to the queue.

Reply:
Not sure if I follow your question.

This change **prevents** from emitting `RequestFulfilled` when repaired the "last slot" (eq. when only one slot failed and was repaired in which case we would get the event, but I am preventing that here).

There is no event emitted when slot is repaired.

Reply:
Ok, when I first reviewed this, I think I thought that the RequestState would change as well when repairing, which obviously does not make sense.

Would it be better to further restrict this to `context.state == RequestState.New`? I'm thinking that if a request has already ended (`RequestState.Finished` or even `Failed`) with one or more of the slots in the `Repair` state, then someone could potentially call `fillSlot` and emit this event (it would, of course incur, a cost, but the "attack" could be disruption of the network).